gammapy/modeling/models/spatial.py

```diff
@@ -154,28 +158,78 @@ def evaluate_geom(self, geom):
         else:
             return self(coords.lon, coords.lat)
 
-    def integrate_geom(self, geom):
+    def integrate_geom(self, geom, oversampling_factor=None):
```

Comment:
Yes for simulations the `SkyModel.integrate_geom` is called directly. The oversampling should present there too.

Reply:
I think the event sampling also relies on `MapEvaluator`, see https://github.com/gammapy/gammapy/blob/master/gammapy/datasets/simulate.py#L70

I completely agree that by API (`integrate_geom()`) it makes completely sense to have the integration implement on the spatial model, however I think the complexity of the code is too high to maintain it independently in two different places. We could go for the compromise here and just keep the "stupid" (i.e. non optimized) integration in `integrate_geom()`. So just do the bare oversampling: no cutouts, no optimized choice of bin sizes etc.

Reply:
The issue is that event sampling does not apply psf convolution and hence relies on `MapEvaluator.complute_flux` which directly uses `SkyModel.integrate_geom`. 
Solving the issue requires to clarify the different evaluation paths in `MapEvaluator` itself. 

I agree that having both at the same time is problematic but sampling with poorly integrated models is problematic as well.
 
I propose we keep as is for now. The correct approach is to ensure that model integration is not duplicated in `MapEvaluator` as well.

If we remove the cutouts, there will be memory issues when oversampling gets too large.

Reply:
Yes, you are right. I missed that we don't rely on `compute_flux()` for the "normal" model evaluation, but have a separate evaluation path for the event sampling. This doesn't make sense. In general this probably shows that the code structure of the `MapEvaluator` together with `SpatialModel.integrate_geom()` is not yet in a good shape. As long as the oversampled case for the event sampling is covered with a unit test, I'm OK to keep as is (maybe leave a TODO...), so it can be simplified / refactored later...